### PR TITLE
feat: generate API Credentials in Admin Portal

### PR DESCRIPTION
**Pull request by yuanyuan-git-tech**
https://2u-internal.atlassian.net/browse/ENT-7415
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/openedx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/openedx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

**Comment by yuanyuan-git-tech** [comment]
A new flag on the Enterprise Customer django admin screen under the subheader “Integration and learning platform settings” called “Allow generation of API credentials”:
Screenshot 2023-08-07 at 9 49 02 AM
Enterprise API endpoints:
http://localhost:18000/enterprise/api/v1/enterprise_customer_api_credentials/
Screenshot 2023-08-07 at 9 50 24 AM

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ jobs:
       - name: Build package
         run: python setup.py sdist bdist_wheel
       - name: Publish to PyPi
-        uses: pypa/gh-action-pypi-publish@master
+        uses: pypa/gh-action-pypi-publish@release/v1
```
Comment (alex-sheehan-edx) [inline .github/workflows/publish.yml:29]: any chance I could get context on this change?

File: tests/test_enterprise/api/test_serializers.py [hunk 2]
```diff
@@ -355,3 +358,47 @@ def test_create(self):
...
+
+        Verify that ``update`` for EnterpriseCustomerAPICredentialsSerializer returns successfully
+        """
+        serializer = EnterpriseCustomerApiCredentialSerializer(self.instance, data=self.data)
+        assert serializer.is_valid()
+        updated_instance = serializer.save()
+        self.assertEqual(updated_instance.name, self.data['name'])
+        self.assertEqual(updated_instance.authorization_grant_type, self.data['authorization_grant_type'])
+        self.assertEqual(updated_instance.client_type, self.data['client_type'])
+        self.assertEqual(updated_instance.redirect_uris, self.data['redirect_uris'])
```
Comment (alex-sheehan-edx) [inline tests/test_enterprise/api/test_serializers.py:404]: I believe you need to add a new line

**Review by alex-sheehan-edx — COMMENTED** [review]
awesome start. A couple of pieces of feed back and I want to sit and think about your tests for a second but we're definitely on our way.

File: enterprise/api/v1/views/enterprise_customer_api_credentials.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        # Verifies the requesting user is connected to an enterprise that has API credentialing bool set to True
+        user = request.user
+        if not has_api_credentials_enabled(user.id):
+            return Response({'detail': 'Can not access to API credentials.'}, status=status.HTTP_405_METHOD_NOT_ALLOWED)
+
+        # Fetches the application for the user
+        # If an application already exists for the user, throw a 409.
+        queryset = self.get_queryset()
+        if queryset:
+            return Response({'detail': 'Application exists.'}, status=status.HTTP_409_CONFLICT)
```
[the remark this answers is not in the document]
Comment (yuanyuan-git-tech) [reply enterprise/api/v1/views/enterprise_customer_api_credentials.py:81]: I saw throw a 409 if an application already exists for the user for the CREATE endpoint in the ticket. Should i generate a new pair of credentials?

File: enterprise/api/utils.py [hunk 4]
```diff
@@ -98,3 +98,26 @@ def get_ent_cust_from_enterprise_customer_key(enterprise_customer_key):
...
     except EnterpriseCustomerInviteKey.DoesNotExist:
         return None
+
+
+def get_enterprise_customer_name_from_user_id(user_id):
+    """
+    Get the enterprise customer name given an user id.
+    """
+    try:
+        return str(EnterpriseCustomerUser.objects.get(user_id=user_id).enterprise_customer.name)
```
[the remark this answers is not in the document]
Comment (yuanyuan-git-tech) [reply enterprise/api/utils.py:108]: Is it that one user corresponds to one EnterpriseCustomerUser? If not, how should i determine the value of Application's name for the current user?

**Review by johnnagro — CHANGES_REQUESTED** [review]
lets huddle tomorrow. i can help you address the merge conflicts and we can discuss a couple feedback items

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @yuanyuan-git-tech!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.


:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://openedx.org/cla)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/openedx/.github/blob/master/CONTRIBUTING.md)     file for more information.     If you've signed an agreement in the past, you may need to re-sign.  See     [The New Home of the Open edX Codebase](https://open.edx.org/blog/the-new-home-of-the-open-edx-codebase/)     for details.

    
Once you've signed the CLA, please allow 1 business day for it to be processed.  After this time, you can re-run the     CLA check by editing the PR title.  If the problem persists, you can tag the `@openedx/cla-problems` team in a     comment on your PR for further assistance.

**Comment by johnnagro** [comment]
attempting to repackage and rework this https://github.com/openedx/edx-enterprise/pull/1854

**Comment by openedx-webhooks** [comment]
@yuanyuan-git-tech Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-enterprise&entry.867055334=https://github.com/openedx/edx-enterprise/pull/1833&entry.1484655318=https://github.com/yuanyuan-git-tech&entry.752974735=2023-08-01+14:20&entry.1917517419=2023-08-29+17:29&entry.2133058324=No) so we can improve your experience in the future.

**Pull request closed without merging**